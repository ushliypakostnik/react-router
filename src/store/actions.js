const fetchUrl = process.env.REACT_APP_API_URL;

export const pageToActive = (page) => ({
  type: "PAGE_TO_ACTIVE",
  activePage: page
});

export const setMinHeight = (minHeight) => ({
  type: "SET_MIN_HEIGHT",
  minHeight: minHeight
});

export const toogleTheme = (theme) => ({
  type: "TOGGLE_THEME",
  theme: theme
});

export const requestAlbums = () => ({
  type: "REQUEST_ALBUMS"
});

export const receiveAlbums = (albums) => ({
  type: "RECEIVE_ALBUMS",
  albums
});

export const requestAlbumFailed = (error) => ({
  type: "REQUEST_ALBUMS_FAILED",
  error
});

export const fetchAlbums = () => {
  return dispatch => {
    dispatch(requestAlbums());
    return fetch(fetchUrl + "/albums")
      .then(res => res.json())
      .then(
        (result) => {
          /* eslint-disable array-callback-return */
          let resultArr = [];
          result.map((item, index) => {
            if (index === 0) {
              resultArr.push({text: item.name, path: "/"});
            } else {
              resultArr.push({text: item.name, path: "/" + item.id});
            }
          });
          dispatch(receiveAlbums(resultArr));
        },
        (error) => {
          dispatch(requestAlbumFailed(error));
        }
      );
  }
}

export const requestData = () => ({
  type: "REQUEST_DATA"
});

export const receiveData = (data) => ({
  type: "RECEIVE_DATA",
  data
});

export const requestDataFailed = (error) => ({
  type: "REQUEST_DATA_FAILED",
  error
});

export const fetchData = (album) => {
  return dispatch => {
    dispatch(requestData());
    return fetch(fetchUrl + "/albums/" + album)
      .then(res => res.json())
      .then(
        (result) => {
          dispatch(receiveData(result));
        },
        (error) => {
          dispatch(requestDataFailed(error));
        }
      );
  }
}
