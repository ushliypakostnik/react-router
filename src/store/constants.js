export const FETCH_URL = process.env.REACT_APP_API_URL;

export const THEME = {
  DARK: "dark",
  LIGHT: "light"
};

export const COOKIES = {
  THEME: "theme"
};
