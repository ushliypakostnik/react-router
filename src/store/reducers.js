const reducer = (state = { }, action) => {
  switch (action.type) {
    case "REQUEST_ALBUMS":
      return Object.assign({}, state, {
        isFetching: true
      });
    case "RECEIVE_ALBUMS":
      return Object.assign({}, state, {
        isFetching: false,
        albums: action.albums
      });
    case "REQUEST_ALBUMS_FAILED":
      return Object.assign({}, state, {
        isFetching: false,
        error: action.error
      });
    case "REQUEST_DATA":
      return Object.assign({}, state, {
        isFetching: true
      });
    case "RECEIVE_DATA":
      return Object.assign({}, state, {
        isFetching: false,
        data: action.data
      });
    case "REQUEST_DATA_FAILED":
      return Object.assign({}, state, {
        isFetching: false,
        error: action.error
      });
    case "PAGE_TO_ACTIVE":
      return Object.assign({}, state, {
        activePage: action.activePage
      });
    case "TOGGLE_THEME":
      return Object.assign({}, state, {
        theme: action.theme
      });
    default:
      return state;
  }
}

export default reducer;
