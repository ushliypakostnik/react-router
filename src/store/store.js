import { createStore, combineReducers, applyMiddleware } from 'redux';
import thunkMiddleware from 'redux-thunk';
import { createLogger } from 'redux-logger';
import { browserHistory } from "react-router";
//import { createBrowserHistory } from "history";
import { syncHistoryWithStore, routerReducer } from 'react-router-redux';

import { INITIAL_STATE } from './constants';
import { pageToActive } from './actions';
import reducer from './reducers';

const middlewares = [];
middlewares.push(thunkMiddleware)

if (process.env.NODE_ENV !== 'production') {
  const loggerMiddleware = createLogger();

  middlewares.push(loggerMiddleware);
}

function configureStore(state) {
  return createStore(
    combineReducers({
      reducer,
      routing: routerReducer
    }),
    state,
    applyMiddleware(...middlewares)
  );
}

const store = configureStore(INITIAL_STATE);

//const history = createBrowserHistory();
export const history = syncHistoryWithStore(browserHistory, store);
//history.listen(location => console.info('-> location:', location));

let lastUrl = "";
if (store.getState().routing.locationBeforeTransitions != null) {
  lastUrl = store.getState().routing.locationBeforeTransitions.pathname;
} else {
  lastUrl = "/";
}

//store.subscribe(() => {
//  console.log("Store: ", store.getState());
//  console.log("last Url: ", lastUrl);
//});

store.dispatch(pageToActive(lastUrl));

export default store;
