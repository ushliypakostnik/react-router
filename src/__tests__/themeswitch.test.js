import React from 'react';
import { Provider } from 'react-redux';
import { CookiesProvider } from 'react-cookie';
import { shallow } from 'enzyme';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import { THEME, INITIAL_STATE } from '../store/constants';

import ThemeSwitch from '../containers/theme-switch';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

describe('theme switch render', () => {

  const props = {
    theme: THEME.DARK
  };

  const state = Object.assign({}, INITIAL_STATE, {
    reducer: {
      theme: props.theme
    }
  });
  const store = mockStore(state);

  const wrapper = mount(<Provider store={store}>
                          <CookiesProvider>
                            <ThemeSwitch {...props} />
                          </CookiesProvider>
                        </Provider>);

  it('theme switch correctly: initial state', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('theme switch wrapper with class', () => {
    expect(wrapper.find('.app__theme-switch.theme-switch')).toHaveLength(1);
  });

  it('theme switch componentDidMount', () => {
    const spy = jest.spyOn(wrapper.find('ThemeSwitch').instance(), 'componentDidMount');
    wrapper.find('ThemeSwitch').instance().componentDidMount();
    expect(spy).toHaveBeenCalled();
  });

  it('theme switch click: set theme', () => {
    const spy = jest.spyOn(wrapper.find('ThemeSwitch').instance(), 'setTheme');
    wrapper.find('a').simulate('click');
    expect(spy).toHaveBeenCalled();
  });

  it('theme switch click: theme toggle', () => {
    const spy = jest.spyOn(wrapper.find('ThemeSwitch').instance(), 'themeChange');
    wrapper.find('a').simulate('click');
    expect(spy).toHaveBeenCalled();
  });

  it('theme switch click: cookies', () => {
    wrapper.find('a').simulate('click');
    expect(wrapper.find('ThemeSwitch').instance().props.cookies.cookies).toEqual({
      theme: THEME.LIGHT
    });
  });

  it('theme switch render light theme', () => {
    const state2 = Object.assign({}, INITIAL_STATE, {
      reducer: {
        theme: THEME.LIGHT
      }
    });
    const store = mockStore(state2);
    const wrapper2 = mount(<Provider store={store}>
                              <ThemeSwitch {...props} />
                           </Provider>);

    expect(wrapper2.find('ThemeSwitch').instance().props.theme).toEqual(THEME.LIGHT);
  });
});
