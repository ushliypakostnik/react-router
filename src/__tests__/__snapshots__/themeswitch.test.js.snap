// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`theme switch render theme switch render correctly 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <CookiesProvider>
    <Connect(Component)
      theme="dark"
    >
      <ForwardRef
        theme="dark"
        toggleTheme={[Function]}
      >
        <withCookies(ThemeSwitch)
          cookies={
            Cookies {
              "HAS_DOCUMENT_COOKIE": true,
              "changeListeners": Array [
                [Function],
              ],
              "cookies": Object {},
            }
          }
          forwardedRef={null}
          theme="dark"
          toggleTheme={[Function]}
        >
          <ThemeSwitch
            allCookies={Object {}}
            cookies={
              Cookies {
                "HAS_DOCUMENT_COOKIE": true,
                "changeListeners": Array [
                  [Function],
                ],
                "cookies": Object {},
              }
            }
            theme="dark"
            toggleTheme={[Function]}
          >
            <a
              aria-label="Theme Switch"
              className="app__theme-switch theme-switch"
              href="#"
              onClick={[Function]}
            >
              <span
                aria-hidden="true"
                className="theme-switch__ico"
              />
              <span
                className="theme-switch__label"
              >
                light
                 theme
              </span>
            </a>
          </ThemeSwitch>
        </withCookies(ThemeSwitch)>
      </ForwardRef>
    </Connect(Component)>
  </CookiesProvider>
</Provider>
`;

exports[`theme switch render theme switch render light theme 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(Component)
    theme="dark"
  >
    <ForwardRef
      theme="light"
      toggleTheme={[Function]}
    >
      <withCookies(ThemeSwitch)
        cookies={
          Cookies {
            "HAS_DOCUMENT_COOKIE": true,
            "changeListeners": Array [
              [Function],
            ],
            "cookies": Object {
              "theme": "light",
            },
          }
        }
        forwardedRef={null}
        theme="light"
        toggleTheme={[Function]}
      >
        <ThemeSwitch
          allCookies={
            Object {
              "theme": "light",
            }
          }
          cookies={
            Cookies {
              "HAS_DOCUMENT_COOKIE": true,
              "changeListeners": Array [
                [Function],
              ],
              "cookies": Object {
                "theme": "light",
              },
            }
          }
          theme="light"
          toggleTheme={[Function]}
        >
          <a
            aria-label="Theme Switch"
            className="app__theme-switch theme-switch"
            href="#"
            onClick={[Function]}
          >
            <span
              aria-hidden="true"
              className="theme-switch__ico"
            />
            <span
              className="theme-switch__label"
            >
              dark
               theme
            </span>
          </a>
        </ThemeSwitch>
      </withCookies(ThemeSwitch)>
    </ForwardRef>
  </Connect(Component)>
</Provider>
`;
